README.md
# @rx-irc/bot-lastfm
Last.fm module for RxBot.

## Installation
1. Clone the [main RxBot repository](https://github.com/rx-irc/bot).
2. Adjust the **package.json** and **config.json** files according to your needs.
3. Install the dependencies with `npm install`.
4. Run the bot with `npm start`.

## License
Copyright Florian Mäder - Permission granted under the [MIT license](LICENSE).


CHANGELOG.md
# [3.0.0](https://github.com/rx-irc/bot-lastfm/compare/v2.1.0...v3.0.0) (2021-06-24)


### Features

* improve logging ([86dbcc6](https://github.com/rx-irc/bot-lastfm/commit/86dbcc6c7d4579cc28db1e612792113eb6ddc51d))
* switch to new client ([356042e](https://github.com/rx-irc/bot-lastfm/commit/356042e895b85da423cfc51727efe28ccc0f964c))


### BREAKING CHANGES

* This needs a different co-dependency.

# [2.1.0](https://github.com/rx-irc/bot-lastfm/compare/v2.0.0...v2.1.0) (2021-05-19)


### Features

* simplify Last.fm API implementation ([4f94251](https://github.com/rx-irc/bot-lastfm/commit/4f942512df5d51baacd1565b70bc28f0c16bea04))

# [2.0.0](https://github.com/rx-irc/bot-lastfm/compare/v1.0.0...v2.0.0) (2021-05-17)


### Features

* switch logging from Winston to debug ([27cf2ed](https://github.com/rx-irc/bot-lastfm/commit/27cf2edb15abbaaeb9502587182127653fbe909d))


### BREAKING CHANGES

* This requires a configuration update.

# 1.0.0 (2021-04-25)


### Features

* semantic-release ([5922bde](https://github.com/rx-irc/bot-lastfm/commit/5922bdeac709cba29cde6cf3fe0d6efa4cc6ae52))


lib/module.js
// NPM Dependencies
const assert = require('assert');
const { filter } = require('rxjs/operators');
const axios = require('axios');

// Local Depdendencies
const logger = require('./logger');
const { version } = require('../package.json');

const REGEXP_BANG = /^!listen (\S+)/i;

let defaults = {};

module.exports = class LastfmModule {
	/**
	 * @param {ClientWrapper} client
	 * @param {object} options
	 * @param {string} options.apiKey
	 */
	constructor(client, options) {
		assert.strictEqual(typeof options.apiKey, 'string');

		/** @type {object} */
		this.settings = { ...defaults, ...options };
		/** @type {string} */
		this.version = version;

		//  ____  _
		// / ___|| |_ _ __ ___  __ _ _ __ ___  ___
		// \___ \| __| '__/ _ \/ _` | '_ ` _ \/ __|
		//  ___) | |_| | |  __/ (_| | | | | | \__ \
		// |____/ \__|_|  \___|\__,_|_| |_| |_|___/
		//

		let listen$ = client.privmsg$.pipe(
			filter(message => REGEXP_BANG.test(message.text))
		);

		//  ____        _                   _       _   _
		// / ___| _   _| |__  ___  ___ _ __(_)_ __ | |_(_) ___  _ __  ___
		// \___ \| | | | '_ \/ __|/ __| '__| | '_ \| __| |/ _ \| '_ \/ __|
		//  ___) | |_| | |_) \__ \ (__| |  | | |_) | |_| | (_) | | | \__ \
		// |____/ \__,_|_.__/|___/\___|_|  |_| .__/ \__|_|\___/|_| |_|___/
		//                                   |_|
		//

		listen$.subscribe(async message => {
			let target = message.target;
			let [, nick] = message.text.match(REGEXP_BANG);

			try {
				let { data } = await axios({
					url: 'https://ws.audioscrobbler.com/2.0/',
					params: {
						api_key: this.settings.apiKey,
						method: 'User.getRecentTracks',
						format: 'json',
						user: nick,
						limit: 1,
					},
				});

				if (data.error) {
					throw new Error(data.message);
				}

				let recent = data.recenttracks;
				let user = recent['@attr']?.user;
				let track = recent.track;

				if (!user) {
					logger.info(`[${nick}] User not found.`);

					client.actionOut$.next({
						command: 'PRIVMSG',
						target,
						text: `${nick} was not found on Last.fm`,
					});
					return;
				}

				if (Array.isArray(track)) {
					track = track[0];
				}

				if (!track || !track.name || !track.artist) {
					logger.info(`[${nick}] Track lookup failed.`);

					client.actionOut$.next({
						command: 'PRIVMSG',
						target,
						text: 'Could not lookup latest track.',
					});
					return;
				}

				let title = track.name;
				let artist = track.artist['#text'];
				let track_string = `${artist} - ${title}`;
				let text = [];

				if (track['@attr']?.['nowplaying'] == 'true') {
					text = [`${user} is listening to:`, track_string];
				} else {
					text = [`${user} last played:`, track_string];
				}

				logger.info(`[${nick}] ${track_string}`);

				client.actionOut$.next({
					command: 'PRIVMSG',
					target,
					text,
				});
			} catch (error) {
				logger.error(`[${nick}] ${error.message}`);

				client.actionOut$.next({
					command: 'PRIVMSG',
					target,
					text: `Error: ${error.message}`,
				});
			}
		});
	}
};


lib/lastfm.js
// Node Dependencies
const http = require('http');

// NPM Depdendencies
const axios = require('axios');
const query = require('qs');

let lastfm = module.exports = {};
let audioscrobbler = { host: 'ws.audioscrobbler.com', port: 80 };

function call(classname, method, options, callback) {
	if (!options.api_key) {
		return callback.call(
			lastfm[classname],
			new Error('API key is missing.'),
			null
		);
	}

	if (!options.format) {
		options.format = 'json';
	}

	let path = '/2.0/?' + query.stringify({
		...options,
		method: classname + '.' + method,
	});

	http.get({ ...audioscrobbler, path }, function (response) {
		let data = '';
		let error = null;

		response.on('data', function (chunk) {
			data += chunk;
		});

		response.on('end', function () {
			// Unknown format let the caller handle the error response
			if (options.format != 'json') {
				callback.call(lastfm[classname], null, data);

				return;
			}

			data = JSON.parse(data);

			if (data.error) {
				error = new Error(data.message);

				error.type = data.error;

				if (error.type == lastfm.ERROR_INVALID_METHOD) {
					throw error;
				}

				callback.call(lastfm[classname], error, null);
			}

			callback.call(lastfm[classname], null, data);
		});
	}).on('error', function (error) {
		callback.call(lastfm[classname], error, null);
	});
}

// ERROR 1 does not exist
lastfm.ERROR_INVALID_SERVICE          =   2
lastfm.ERROR_INVALID_METHOD           =   3
lastfm.ERROR_AUTHENTICATION_FAILED    =   4
lastfm.ERROR_INVALID_FORMAT           =   5
lastfm.ERROR_INVALID_PARAMETERS       =   6
lastfm.ERROR_INVALID_RESOURCE         =   7
lastfm.ERROR_OPERATION_FAILED         =   8
lastfm.ERROR_INVALID_SESSION_KEY      =   9
lastfm.ERROR_INVALID_API_KEY          =  10
lastfm.ERROR_SERVICE_OFFLINE          =  11
lastfm.ERROR_SUBSCRIBERS_ONLY         =  12
lastfm.ERROR_INVALID_METHOD_SIGNATURE =  13
lastfm.ERROR_UNAUTHORIZED_TOKEN       =  14
lastfm.ERROR_ITEM_NOT_STREAMABLE      =  15
lastfm.ERROR_SERVICE_UNAVAILABLE      =  16
lastfm.ERROR_NOT_LOGGED_IN            =  17
lastfm.ERROR_TRIAL_EXPIRED            =  18
// ERROR 19 does not exist
lastfm.ERROR_NOT_ENOUGH_CONTENT       =  20
lastfm.ERROR_NOT_ENOUGH_MEMBERS       =  21
lastfm.ERROR_NOT_ENOUGH_FANS          =  22
lastfm.ERROR_NOT_ENOUGH_NEIGHBOURS    =  23
lastfm.ERROR_NO_PEAK_RADIO            =  24
lastfm.ERROR_RADIO_NOT_FOUND          =  25
lastfm.ERROR_API_KEY_SUSPENDED        =  26
lastfm.ERROR_DEPRECATED_REQUEST       =  27
lastfm.ERROR_RATE_LIMIT_EXCEEDED      =  28

let classes = [
	'Album',
	'Artist',
	'Auth',
	'Chart',
	'Event',
	'Geo',
	'Group',
	'Library',
	'Playlist',
	'Radio',
	'Tag',
	'Tasteometer',
	'Track',
	'User',
	'Venue',
];

for (let classname of classes) {
	lastfm[classname] = new Proxy({}, {
		get: function (proxy, property) {
			return function () {
				return call(classname, property, arguments[0], arguments[1]);
			};
		},
	});
}
